### ISSUE TITLE
DatasetBuilder - create SFT/DPO training pairs from traces

### ISSUE THREAD

**Issue by boersmamarcel** [boersmamarcel/chatty2#165]
## Context

Raw traces from TraceStore aren't directly consumable by the training loop. DatasetBuilder converts structured trace data into properly formatted, tokenized, and batched training pairs ready for Candle consumption.

### Core Struct

```rust
pub struct DatasetBuilder {
    tokenizer: tokenizers::Tokenizer,
    config: DatasetConfig,
}

pub struct DatasetConfig {
    pub max_seq_length: usize,          // Model's context window (e.g., 4096)
    pub min_response_tokens: usize,     // Skip very short responses
    pub train_split: f32,               // e.g., 0.9 (90% train, 10% val)
    pub dedup_threshold: f32,           // Cosine similarity threshold for dedup
    pub batch_size: usize,              // Training batch size
}
```

### SFT Dataset

```rust
pub struct SftDataset {
    pub train: Vec<SftSample>,
    pub validation: Vec<SftSample>,
}

pub struct SftSample {
    pub input_ids: Vec<u32>,           // Tokenized full sequence
    pub labels: Vec<i64>,             // -100 for prompt tokens, token IDs for response
    pub attention_mask: Vec<u32>,
}
```

Building from traces:
1. Query TraceStore for positive-feedback traces
2. Format as chat template: `<|user|>...<|assistant|>...`
3. Tokenize with model's tokenizer
4. Create labels (mask prompt tokens with -100, keep response tokens)
5. Truncate/pad to `max_seq_length`
6. Split into train/validation

### DPO Dataset

```rust
pub struct DpoDataset {
    pub train: Vec<DpoSample>,
    pub validation: Vec<DpoSample>,
}

pub struct DpoSample {
    pub prompt_ids: Vec<u32>,          // Tokenized conversation context
    pub chosen_ids: Vec<u32>,          // Tokenized preferred response
    pub rejected_ids: Vec<u32>,        // Tokenized rejected response
}
```

Building from regeneration records:
1. Query TraceStore for regeneration pairs
2. Build prompt from conversation context up to the regenerated turn
3. Tokenize prompt, chosen (regenerated), and rejected (original) separately
4. Apply length constraints

### Deduplication

Near-duplicate detection to avoid training on repetitive data:
- Hash-based exact dedup (fast, catches identical messages)
- Optional: token-overlap ratio for near-dedup

### Pipeline

```rust
impl DatasetBuilder {
    pub fn build_sft_dataset(
        &self,
        trace_store: &TraceStore,
        filter: &TraceFilter,
    ) -> Result<SftDataset>;

    pub fn build_dpo_dataset(
        &self,
        trace_store: &TraceStore,
        filter: &TraceFilter,
    ) -> Result<DpoDataset>;

    pub fn dataset_stats(&self, dataset: &SftDataset) -> DatasetStats;
}

pub struct DatasetStats {
    pub total_examples: usize,
    pub train_examples: usize,
    pub validation_examples: usize,
    pub avg_sequence_length: f32,
    pub total_tokens: usize,
    pub duplicates_removed: usize,
}
```

## Acceptance Criteria

- [ ] `build_sft_dataset()` produces tokenized, label-masked training samples
- [ ] `build_dpo_dataset()` produces prompt/chosen/rejected triples
- [ ] Proper chat template formatting for target model architectures
- [ ] Label masking: prompt tokens masked with -100, only response tokens used for loss
- [ ] Train/validation split with configurable ratio
- [ ] Sequence length truncation and padding
- [ ] Deduplication (exact match at minimum)
- [ ] `dataset_stats()` provides useful diagnostics
- [ ] Works with different tokenizers (Phi, Gemma, Qwen)
- [ ] Unit tests with synthetic trace data

## Depends On
- #163 (CandleLocalModel — tokenizer and model architecture info)
- #164 (TraceStore — data source)

## Blocks
- #155 Issue 11 (LoRA training loop consumes datasets)
